ocaml/sdk-gen/java/autogen/xen-api/src/main/java/com/xensource/xenapi/Connection.java

```diff
@@ -29,228 +29,211 @@
+     * @param connectionTimeout The connection timeout for JSON-RPC calls in seconds
      */
-    public Connection(URL url, String sessionReference, int replyTimeout, int connTimeout)
-    {
-        this.client = getClientFromURL(url, replyTimeout, connTimeout);
+    public Connection(URL url, String sessionReference, int requestTimeout, int connectionTimeout) {
+        this.client = new JsonRpcClient(url, requestTimeout, connectionTimeout);
         this.sessionReference = sessionReference;
     }
 
-    private XmlRpcClientConfigImpl config = new XmlRpcClientConfigImpl();
-
-    public XmlRpcClientConfigImpl getConfig()
-    {
-        return config;
+    /**
+     * Updated when Session.login_with_password() is called.
+     */
+    public APIVersion getAPIVersion() {
+        return apiVersion;
     }
 
-    private XmlRpcClient getClientFromURL(URL url, int replyWait, int connWait)
-    {
-        config.setTimeZone(TimeZone.getTimeZone("UTC"));
-        config.setServerURL(url);
-        config.setReplyTimeout(replyWait * 1000);
-        config.setConnectionTimeout(connWait * 1000);
-        XmlRpcClient client = new XmlRpcClient();
-        client.setConfig(config);
-        return client;
+    private void setAPIVersion(Session session) throws IOException {
+        try {
+            long major = session.getThisHost(this).getAPIVersionMajor(this);
+            long minor = session.getThisHost(this).getAPIVersionMajor(this);
+            apiVersion = APIVersion.fromMajorMinor(major, minor);
+        } catch (BadServerResponse exn) {
+            apiVersion = APIVersion.UNKNOWN;
+        }
     }
 
     /*
      * Because the binding calls are constructing their own parameter lists, they need to be able to get to
      * the session reference directly. This is all rather ugly and needs redone
-     * Changed to public to allow easier integration with HTTP-level streaming interface,
-     * see CA-15447
+     * CA-15447: Changed to public in order to allow easier integration with HTTP-level streaming interface,
      */
-    public String getSessionReference()
-    {
+    public String getSessionReference() {
         return this.sessionReference;
     }
 
     /**
-     * The (auto-generated parts of) the bindings dispatch XMLRPC calls on this Connection's client through this method.
+     * Send a method call to xapi's backend. You need to provide the type of the data returned by a successful response.
+     *
+     * @param methodCall            the JSON-RPC xapi method call. e.g.: session.login_with_password
+     * @param methodParameters      the methodParameters of the method call
+     * @param responseTypeReference the type of the response, wrapped with a TypeReference
+     * @param <T>                   The type of the response's payload. For instance, a map of opaque references to VM objects is expected when calling VM.get_all_records
+     * @return The result of the call with the type specified under T.
```

Comment:
If you're making other changes you may want to fix the indentation of the parameters in both dispatch methods.
(Apropos, the indentation is off in some of the autogeenrated files, but no need to bother with those in this PR as they can be fixed when we switch to templates).